### ISSUE TITLE
Auto-indent should be a single undo operation

### ISSUE THREAD

**Issue by Gama11** [fdorg/flashdevelop#696]
Pressing Enter here is a single input but takes two steps to undo, which is annoying:

![autoindentundo](https://cloud.githubusercontent.com/assets/2620907/7784076/bf47ea00-0158-11e5-8884-9b06e1eee10f.gif)

**Comment by elsassph** [comment]
The rationale is that it allows you to undo the auto-indent operation without undoing the newline, otherwise that would be a new category of annoyance...

**Comment by Gama11** [comment]
I don't agree with that logic... There shouldn't be a discrepancy between "do" and "undo" - an action that required a single input from the user should also only require a single action to undo.

You can test this in IntelliJ IDEA, Visual Studio, Eclipse or even Sublime Text - it's a single undo action everywhere.

I wouldn't even have considered using undo to undo auto-indent as a use-case - I would simply press backspace to delete the indent. I'm pretty sure most people don't even think of auto-indent as a feature, for them it's just "something that IDEs do".

**Comment by elsassph** [comment]
Well the rationale isn't always very clever... ;) We should try fixing this.

**Comment by Gama11** [comment]
That's actually proving somewhat tricky, since smart indent is handled in a CharAdded event handler - the BeginUndoAction() call would need to be placed before the insertion of the newline character, but I don't think Scintilla has a BeforeCharAdded event.

**Comment by Marc010** [comment]
I just tested it and in Sublime Text Build 3083 I need to press backspace 3 times in this situation to undo the enter keypress. But it works in IntelliJ and i think it would be a useful addition to FD :)

**Comment by Gama11** [comment]
@Marc010 we're talking about undo / Ctrl + Z, not backspace.

**Comment by Gama11** [comment]
I've wanted to take a look at Notepad++ to see how they do it in Scintilla, turns out it behaves the same way as FD currently.

**Comment by elsassph** [comment]
It sounds like it requires to capture the Enter key and to process in a completely custom way.

**Comment by Neverbirth** [comment]
Actually, I have some work done that would allow for this. It isn't committed because FD 5 is feature frozen.

Also, it brings one question to the table... should we add another event for getting the key before it's sent, or should we change all the code to get the keys before they are sent? of course the latter means rewriting a lot of code, but I think the question is interesting, because it's the standard behaviour when working with control frameworks.
